functions/src/update-stats.ts

```diff
@@ -0,0 +1,70 @@
+import * as functions from "firebase-functions";
+import * as firebase from "firebase-admin";
+
+import { Config, isEnabled, asBoolean } from "./config";
+import { Vote } from "./cast-vote";
+
+const config = functions.config() as Config;
+
+const validVotes = ["great", "notThatGreat", "notGreatAtAll"];
+
+interface StatsData {
+  greatCount: number;
+  notThatGreatCount: number;
+  notGreatAtAllCount: number;
+}
+
+//updateStats({value: "great", voter: "example@zenika.com", campaign: "2018-09", recordedAt: "2018-09-19T15:30:00.00Z", agency: "Nantes"})
+
+export const updateStats = functions.firestore
+  .document("vote/{voteId}")
+  .onCreate(async voteSnapshot => {
+    if (!isEnabled(config.features.collect_stats)) {
+      console.info("stats collecting is disabled; aborting");
+      return;
+    }
+    const vote = voteSnapshot.data()! as Vote;
+    let stats: StatsData;
+    await firebase.firestore().runTransaction(async transaction => {
+      await transaction
+        .get(
+          firebase
+            .firestore()
+            .collection("stats")
+            .doc(vote.campaign)
+        )
+        .then(doc => {
+          stats = doc.data()! as StatsData;
+          if (!stats) {
+            stats = {
+              greatCount: 0,
+              notThatGreatCount: 0,
+              notGreatAtAllCount: 0
+            };
+          }
+          switch (vote.value) {
+            case validVotes[0]:
+              stats.greatCount += 1;
+              break;
+            case validVotes[1]:
+              stats.notThatGreatCount += 1;
+              break;
+            case validVotes[2]:
+              stats.notGreatAtAllCount += 1;
+              break;
+          }
```

Comment:
This works great, but it has a subtle implication. This code works only if the value of the vote is one of the 3 accepted values. It knows about those 3 valid values. It has the knowledge of them. As a consumer of votes, it is coupled (tightly linked) to the producer of votes (the cast vote function). Coupling makes the code harder to change.

For example, let's say in the future we want to extend the number of possible vote values to 4. In that case, we have to come and change this code in addition to the code in cast vote. This is a bit unfortunate because we want to prevent mistakes like this.

What can we do then?

Well, the stats feature goal is to count how many of the same vote value, for all vote values. It does not care about what those values are. If 2 values are equal, it counts them together, if there are not, it counts them separately. For it to know the values would be over-specifying.

Can you try to refactor the code to make it do that? :)